src/logic/div.rs

```diff
@@ -0,0 +1,1290 @@
+                    .wrapping_add($add[$len-1])
+                    .wrapping_add(carry);
+            }}
+        }
+
+        //assumes that:
+        // - ini_duo_sd > 1
+        // - div_sd > 1
+        // - (`duo` / `div`) > 1
+        #[inline(always)]
+        fn large_div_by_large(
+            duo: &mut [Digit], //the dividend which will become the quotient
+            ini_duo_sd: usize, //the initial most significant digit of `duo`
+            ini_duo_sb: usize, //the number of significant bits in `duo`
+            ini_duo_lz: usize, //the number of leading zeros in `duo[ini_duo_sd]`
+            div: &mut [Digit], //the divisor which will become the remainder
+            div_sd: usize, //the most significant digit of `div`
+            div_sb: usize, //the number of significant bits in `div`
+            div_lz: usize //the number of leading zeros in `div[div_sd]`
+        ) {
+            //difference between the places of the most significant bits
+            let ini_diff_bits = ini_duo_sb - div_sb;
+            if ini_diff_bits < Digit::BITS {
+                //the `mul` or `mul - 1` algorithm
+                let (duo_sig_dd, div_sig_dd) = if ini_duo_lz == 0 {
+                    //avoid shr overflow
+                    (
+                        DoubleDigit::from_lo_hi(duo[ini_duo_sd - 1], duo[ini_duo_sd]),
+                        DoubleDigit::from_lo_hi(div[ini_duo_sd - 1], div[ini_duo_sd]) 
+                    )
+                } else {
+                    (
+                        (duo[ini_duo_sd].dd() << (ini_duo_lz + Digit::BITS)) |
+                        (duo[ini_duo_sd - 1].dd() << ini_duo_lz) |
+                        (duo[ini_duo_sd - 2].dd() >> (Digit::BITS - ini_duo_lz)),
+                        (div[ini_duo_sd].dd() << (ini_duo_lz + Digit::BITS)) |
+                        (div[ini_duo_sd - 1].dd() << ini_duo_lz) |
+                        (div[ini_duo_sd - 2].dd() >> (Digit::BITS - ini_duo_lz))
+                    )
+                };
+                let mul = duo_sig_dd.wrapping_div(div_sig_dd).lo();
+                //Allocation could be avoided, but if it turns out `mul - 1` should be used, more
+                //long division would have to occur to recover `div`, followed by a second long
+                //multiplication with `mul - 1`.
+                //this will become `-(div * mul)`
+                let mut sub: Vec<Digit> = Vec::with_capacity(ini_duo_sd + 1);
+                let mut carry = Digit::zero();
+                for i in 0..=ini_duo_sd {
+                    let tmp = mul.carrying_mul_add(div[i], carry);
+                    sub.push(tmp.0);
+                    carry = tmp.1;
+                }
+                let sub_len = sub.len();
+                //when `div * mul > duo`, sometimes `sub` can overflow to a digit higher than the
+                //digits availiable in the slice, which has to be handled in this way for max
+                //performance
```

Comment:
Please describe more in-depth about the rationals and reasons behind this performance block of code. Readers are curious. ;)